File: modules/project/src/main/java/com/falsepattern/zigbrains/project/ide/project/ZigProjectSettingsPanel.java
/*
 * Copyright 2023-2024 FalsePattern
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.falsepattern.zigbrains.project.ide.project;

import com.falsepattern.zigbrains.common.util.PathUtil;
import com.falsepattern.zigbrains.common.util.StringUtil;
import com.falsepattern.zigbrains.common.util.TextFieldUtil;
import com.falsepattern.zigbrains.common.util.dsl.JavaPanel;
import com.falsepattern.zigbrains.project.openapi.MyDisposable;
import com.falsepattern.zigbrains.project.openapi.UIDebouncer;
import com.falsepattern.zigbrains.project.openapi.components.ZigProjectSettings;
import com.falsepattern.zigbrains.project.openapi.components.ZigProjectSettingsService;
import com.falsepattern.zigbrains.project.toolchain.AbstractZigToolchain;
import com.falsepattern.zigbrains.project.toolchain.ZigToolchainProvider;
import com.falsepattern.zigbrains.project.toolchain.tools.ZigCompilerTool;
import com.intellij.openapi.project.ProjectManager;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import com.intellij.openapi.util.Disposer;
import com.intellij.openapi.util.Pair;
import com.intellij.ui.JBColor;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.dsl.builder.AlignX;
import lombok.Getter;
import lombok.val;

import javax.swing.JLabel;
import java.awt.event.ActionEvent;
import java.util.Optional;

import static com.falsepattern.zigbrains.common.util.KtUtil.$f;

public class ZigProjectSettingsPanel implements MyDisposable {
    @Getter
    private boolean disposed = false;

    private final UIDebouncer versionUpdateDebouncer = new UIDebouncer(this);

    private final TextFieldWithBrowseButton pathToToolchain = TextFieldUtil.pathToDirectoryTextField(this,
                                                                                                     "Path to the Zig Toolchain",
                                                                                                     this::updateUI);

    private final JLabel toolchainVersion = new JLabel();

    private final JBCheckBox stdFieldOverride = new JBCheckBox("Override");

    private final TextFieldWithBrowseButton pathToStdField = TextFieldUtil.pathToDirectoryTextField(this,
                                                                                                    "Path to Standard Library",
                                                                                                    () -> {});

    {
        stdFieldOverride.addChangeListener(e -> {
            if (stdFieldOverride.isSelected()) {
                pathToStdField.setEnabled(true);
            } else {
                pathToStdField.setEnabled(false);
                updateUI();
            }
        });
    }

    private void autodetect(ActionEvent e) {
        autodetect();
    }

    public void autodetect() {
        val tc = AbstractZigToolchain.suggest();
        if (tc != null) {
            pathToToolchain.setText(PathUtil.stringFromPath(tc.getLocation()));
            updateUI();
        }
    }

    public ZigProjectSettings getData() {
        val toolchain = Optional.of(pathToToolchain.getText())
                                .map(PathUtil::pathFromString)
                                .map(ZigToolchainProvider::findToolchain)
                                .orElse(null);
        return new ZigProjectSettings(stdFieldOverride.isSelected() ? StringUtil.blankToNull(pathToStdField.getText()) : null, toolchain);
    }

    public void setData(ZigProjectSettings value) {
        pathToToolchain.setText(Optional.ofNullable(value.getToolchainHomeDirectory())
                                        .orElse(""));

        stdFieldOverride.setSelected(value.overrideStdPath);

        pathToStdField.setText(Optional.ofNullable(value.getExplicitPathToStd()).orElse(""));

        pathToStdField.setEnabled(value.overrideStdPath);

        updateUI();
    }

    public void attachPanelTo(JavaPanel p) {
        Optional.ofNullable(ZigProjectSettingsService.getInstance(ProjectManager.getInstance().getDefaultProject()))
                .map(ZigProjectSettingsService::getState)
                .ifPresent(this::setData);
        p.group("Zig Settings", true, p2 -> {
            p2.row("Toolchain location", r -> {
                r.cell(pathToToolchain).resizableColumn().align(AlignX.FILL);
                r.button("Autodetect", $f(this::autodetect));
            });
            p2.cell("Toolchain version", toolchainVersion);
            p2.cell("Override standard library path", stdFieldOverride);
            p2.row("Standard library location", row -> {
                row.cell(pathToStdField).resizableColumn().align(AlignX.FILL);
                row.cell(stdFieldOverride);
            });
        });
    }

    @Override
    public void dispose() {
        disposed = true;
        Disposer.dispose(pathToToolchain);
    }

    private void updateUI() {
        val pathToToolchain = PathUtil.pathFromString(this.pathToToolchain.getText());

        versionUpdateDebouncer.run(
                () -> {
                    val toolchain = Optional.ofNullable(pathToToolchain).map(ZigToolchainProvider::findToolchain).orElse(null);
                    val zig = Optional.ofNullable(toolchain).map(AbstractZigToolchain::zig).orElse(null);
                    val version = Optional.ofNullable(zig).flatMap(ZigCompilerTool::queryVersion).orElse(null);
                    val stdPath = Optional.ofNullable(zig).flatMap(ZigCompilerTool::getStdPath).orElse(null);

                    return new Pair<>(version, stdPath);
                    },
                (pair) -> {
                    val zigVersion = pair.first;
                    val stdPath = pair.second;
                    toolchainVersion.setText(StringUtil.orEmpty(zigVersion));
                    toolchainVersion.setForeground(JBColor.foreground());

                    if (!stdFieldOverride.isSelected())
                        pathToStdField.setText(StringUtil.orEmpty(stdPath));
                });
    }
}

File: modules/core/src/main/kotlin/com/falsepattern/zigbrains/project/settings/ZigProjectSettingsPanel.kt
/*
 * This file is part of ZigBrains.
 *
 * Copyright (C) 2023-2024 FalsePattern
 * All Rights Reserved
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * ZigBrains is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, only version 3 of the License.
 *
 * ZigBrains is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ZigBrains. If not, see <https://www.gnu.org/licenses/>.
 */

package com.falsepattern.zigbrains.project.settings

import com.falsepattern.zigbrains.ZigBrainsBundle
import com.falsepattern.zigbrains.direnv.DirenvCmd
import com.falsepattern.zigbrains.project.toolchain.LocalZigToolchain
import com.falsepattern.zigbrains.project.toolchain.ZigToolchainProvider
import com.falsepattern.zigbrains.shared.coroutine.launchWithEDT
import com.falsepattern.zigbrains.shared.coroutine.withEDTContext
import com.falsepattern.zigbrains.shared.zigCoroutineScope
import com.intellij.openapi.Disposable
import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory
import com.intellij.openapi.project.Project
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.util.UserDataHolderBase
import com.intellij.openapi.util.io.toNioPathOrNull
import com.intellij.platform.ide.progress.ModalTaskOwner
import com.intellij.platform.ide.progress.TaskCancellation
import com.intellij.platform.ide.progress.withModalProgress
import com.intellij.ui.DocumentAdapter
import com.intellij.ui.JBColor
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.textFieldWithBrowseButton
import com.intellij.ui.dsl.builder.AlignX
import com.intellij.ui.dsl.builder.Panel
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import javax.swing.event.DocumentEvent
import kotlin.io.path.pathString

class ZigProjectSettingsPanel(private val project: Project?) : Disposable {
    private val direnv = JBCheckBox(ZigBrainsBundle.message("settings.project.label.direnv"))
    private val pathToToolchain = textFieldWithBrowseButton(
        project,
        ZigBrainsBundle.message("dialog.title.zig-toolchain"),
        FileChooserDescriptorFactory.createSingleFolderDescriptor()
    ).also {
        it.textField.document.addDocumentListener(object : DocumentAdapter() {
            override fun textChanged(e: DocumentEvent) {
                dispatchUpdateUI()
            }
        })
        Disposer.register(this, it)
    }
    private val toolchainVersion = JBLabel()
    private val stdFieldOverride = JBCheckBox(ZigBrainsBundle.message("settings.project.label.override-std")).apply {
        addChangeListener {
            if (isSelected) {
                pathToStd.isEnabled = true
            } else {
                pathToStd.isEnabled = false
                updateUI()
            }
        }
    }
    private val pathToStd = textFieldWithBrowseButton(
        project,
        ZigBrainsBundle.message("dialog.title.zig-std"),
        FileChooserDescriptorFactory.createSingleFolderDescriptor()
    ).also { Disposer.register(this, it) }
    private var debounce: Job? = null

    suspend fun autodetect() {
        val data = UserDataHolderBase()
        data.putUserData(LocalZigToolchain.DIRENV_KEY, direnv.isSelected)
        val tc = ZigToolchainProvider.suggestToolchain(project, data) ?: return
        if (tc !is LocalZigToolchain) {
            TODO("Implement non-local zig toolchain in config")
        }
        pathToToolchain.text = tc.location.pathString
        dispatchUpdateUI()
    }

    var data
        get() = ZigProjectSettings(
            direnv.isSelected,
            stdFieldOverride.isSelected,
            pathToStd.text,
            pathToToolchain.text
        )
        set(value) {
            direnv.isSelected = value.direnv
            pathToToolchain.text = value.toolchainPath ?: ""
            stdFieldOverride.isSelected = value.overrideStdPath
            pathToStd.text = value.explicitPathToStd ?: ""
            pathToStd.isEnabled = value.overrideStdPath
            dispatchUpdateUI()
        }

    fun attach(p: Panel): Unit = with(p) {
        val project = project ?: ProjectManager.getInstance().defaultProject
        data = project.zigProjectSettings.state
        group(ZigBrainsBundle.message("settings.project.group.title")) {
            row(ZigBrainsBundle.message("settings.project.label.toolchain")) {
                cell(pathToToolchain).resizableColumn().align(AlignX.FILL)
                if (DirenvCmd.direnvInstalled() && !project.isDefault) {
                    cell(direnv)
                }
                button(ZigBrainsBundle.message("settings.project.label.toolchain-autodetect")) {
                    project.zigCoroutineScope.launchWithEDT {
                        withModalProgress(ModalTaskOwner.component(pathToToolchain), "Detecting Zig...", TaskCancellation.cancellable()) {
                            autodetect()
                        }
                    }
                }
            }
            row(ZigBrainsBundle.message("settings.project.label.toolchain-version")) {
                cell(toolchainVersion)
            }
            row(ZigBrainsBundle.message("settings.project.label.std-location")) {
                cell(pathToStd).resizableColumn().align(AlignX.FILL)
                cell(stdFieldOverride)
            }
        }
    }

    private fun dispatchUpdateUI() {
        debounce?.cancel("New debounce")
        debounce = project.zigCoroutineScope.launch {
            updateUI()
        }
    }


    private suspend fun updateUI() {
        val pathToToolchain = this.pathToToolchain.text.toNioPathOrNull()
        delay(200)
        val toolchain = pathToToolchain?.let { LocalZigToolchain(it) }
        val zig = toolchain?.zig
        val env = zig?.getEnv(project)
        val version = env?.version
        val stdPath = env?.stdPath(toolchain)
        withEDTContext {
            toolchainVersion.text = version ?: ""
            toolchainVersion.foreground = JBColor.foreground()

            if (!stdFieldOverride.isSelected) {
                pathToStd.text = stdPath?.pathString ?: ""
            }
        }
    }

    override fun dispose() {
        debounce?.cancel("Disposed")
    }
}